Reject empty path string when parsing offset-path and clip-path

darinadler opened the pull request:
#### daaddb2470406dd65ba77d155c03b0e6eb60c047
<pre>
Reject empty path string when parsing offset-path and clip-path
<a href="https://bugs.webkit.org/show_bug.cgi?id=265666">https://bugs.webkit.org/show_bug.cgi?id=265666</a>
<a href="https://rdar.apple.com/119034794">rdar://119034794</a>

Reviewed by Tim Nguyen.

Reject empty paths. Reject fill rules for offset-path.

This is what the WPT tests expect, although I can&apos;t find any specification that clearly
asks for either of these behaviors. It&apos;s expedient for now to match the tests and the
other browsers, although it would be better if the specification matched.

* LayoutTests/imported/w3c/web-platform-tests/css/css-masking/parsing/clip-path-invalid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/motion/parsing/offset-path-parsing-invalid-expected.txt:
Expect more PASS.

* Source/WebCore/css/CSSProperties.json: Changed parameters for consumePathOperation
from ConsumeRay::Exclude/Include to RejectRay and RejectFillRule.

* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumeBasicShapePath): Added support for
RejectFillRule. Reject empty paths.
(WebCore::CSSPropertyParserHelpers::consumeBasicShape): Added options passed
through to consumeBasicShapePath.
(WebCore::CSSPropertyParserHelpers::consumeBasicShapeRayOrBox): Take options
instead of the consumeRay argument. Pass the options through to consumeBasicShape,
and implement the RejectRay option.
(WebCore::CSSPropertyParserHelpers::consumePathOperation): Take options instead
of the consumeRay argument.
(WebCore::CSSPropertyParserHelpers::consumeShapeOutside): Pass options to consumeBasicShape.

* Source/WebCore/css/parser/CSSPropertyParserHelpers.h: Added PathParsingOption, removed
PathParsingOption. Made consumePathOperation take options instead of the consumeRay argument.

Canonical link: <a href="https://commits.webkit.org/271788@main">https://commits.webkit.org/271788@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c22428796952af649ef277bff6efbb4743efc5da

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/29590 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/8250 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/30898 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/32101 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/26796 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/30193 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/10401 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/5539 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/26808 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/29863 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/6889 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/25263 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/5884 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/6029 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/26351 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/33445 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/26987 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/26772 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/32247 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/5974 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/4164 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/30021 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/7709 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7040 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/6515 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/6497 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
```diff
@@ -6935,6 +6937,8 @@ static RefPtr<CSSPathValue> consumeBasicShapePath(CSSParserTokenRange& args)
     if (!buildSVGPathByteStreamFromString(args.consumeIncludingWhitespace().value(), byteStream, UnalteredParsing))
         return nullptr;
+    if (options.contains(RejectEmptyPath) && byteStream.isEmpty())
+        return nullptr;
```
I think it would make sense to always ignore the empty paths including for the `shape-outside` property, which you've excluded here. So I would remove the option, and unconditionally ignore.

We should add a WPT for clip-path / shape-outside too.

darinadler replied:
The test for `clip-path` does exist, it’s `css/css-masking/parsing/clip-path-invalid.html`.

I think when I did this unconditionally, some existing tests failed. I will look into `shape-outside` separately and do what you suggest if it works.

darinadler replied:
I don’t understand the status of `shape-outside`. The `css/css-shapes/parsing/shape-outside-valid.html` and `css/css-shapes/parsing/shape-outside-invalid.html` do not include any cases of `path()`.

darinadler committed:
Reject empty path string when parsing offset-path and clip-path
https://bugs.webkit.org/show_bug.cgi?id=265666
rdar://119034794

Reviewed by Tim Nguyen.

Reject empty paths. Reject fill rules for offset-path.

This is what the WPT tests expect, although I can't find any specification that clearly
asks for either of these behaviors. It's expedient for now to match the tests and the
other browsers, although it would be better if the specification matched.

* LayoutTests/imported/w3c/web-platform-tests/css/css-masking/parsing/clip-path-invalid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/motion/parsing/offset-path-parsing-invalid-expected.txt:
Expect more PASS.

* Source/WebCore/css/CSSProperties.json: Changed parameters for consumePathOperation
from ConsumeRay::Exclude/Include to RejectRay and RejectFillRule.

* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumeBasicShapePath): Added support for
RejectFillRule. Reject empty paths.
(WebCore::CSSPropertyParserHelpers::consumeBasicShape): Added options passed
through to consumeBasicShapePath.
(WebCore::CSSPropertyParserHelpers::consumeBasicShapeRayOrBox): Take options
instead of the consumeRay argument. Pass the options through to consumeBasicShape,
and implement the RejectRay option.
(WebCore::CSSPropertyParserHelpers::consumePathOperation): Take options instead
of the consumeRay argument.
(WebCore::CSSPropertyParserHelpers::consumeShapeOutside): Pass options to consumeBasicShape.

* Source/WebCore/css/parser/CSSPropertyParserHelpers.h: Added PathParsingOption, removed
PathParsingOption. Made consumePathOperation take options instead of the consumeRay argument.

Canonical link: https://commits.webkit.org/271788@main
